### PR TITLE
Theme selector

### LINKED ISSUE #5: Theme switching
Add a toggle to switch between dark and light mode

### PR DESCRIPTION
**Pull request by Lemonatix**
First attempt of adding a working theme-switch for the app.

### CODE REVIEW

File: theme-switch/main-theme-switch (V1.0).dart [hunk 1]
```diff
@@ -0,0 +1,60 @@
+// theme-switch (1.0) as .dart file
+
+import 'package:flutter/material.dart';
+import 'package:shared_preferences/shared_preferences.dart'
+
+void main() {
+    runApp(MyApp());
+}
+
+// MyApp = stateful widget, that maintains current theme state
```
Comment (biosfood) [inline theme-switch/main-theme-switch (V1.0).dart]: remove this, this seems clear

File: theme-switch/main-theme-switch (V1.0).dart [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+class _MyAppState extends State<MyApp> {
+    bool _isDarkMode = false;
+
+    @override
+    void initState() { // initState Method: loads the theme preference from shared preference, when App starts
+        super.initState();
+        _loadTheme();
+    }
```
Comment (biosfood) [inline theme-switch/main-theme-switch (V1.0).dart]: spacing

File: theme-switch/main-theme-switch (V1.0).dart [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            ),
+            body: Center(
+                child: SwitchListTile( // Allows user to toggle theme, which updates the state and saves the preference
+                    title: Text('Dark Mode'),
+                    value: _isDarkMode,
+                    onChanged: (value) {
+                        setState(() {
+                            _isDarkMode = value;
+                            _saveTheme(value);
+                        }
```
Comment (biosfood) [inline theme-switch/main-theme-switch (V1.0).dart]: Indentation and semicolon, or remove brackets

File: theme-switch/main-theme-switch (V1.0).dart [hunk 4]
```diff
@@ -0,0 +1,60 @@
+// theme-switch (1.0) as .dart file
```
Comment (biosfood) [inline theme-switch/main-theme-switch (V1.0).dart]: in lib folder, maybe UI or settings folder
remove 1.0 from file name and comment
remove comment

File: theme-switch/main-theme-switch (V1.0).dart [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+import 'package:flutter/material.dart';
+import 'package:shared_preferences/shared_preferences.dart'
+
+void main() {
```
Comment (biosfood) [inline theme-switch/main-theme-switch (V1.0).dart]: no main function other than in main.dart or libs

File: theme-switch/theme-switch-dependences.yml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# Add Dependencies
+# shared_preferences package must be included in pubspec.yaml (file for saving user's preference)
+
+dependencies:
```
Comment (biosfood) [inline theme-switch/theme-switch-dependences.yml]: not in a separate file

File: theme-switch/main-theme-switch (V1.0).dart [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+class _MyAppState extends State<MyApp> {
+    bool _isDarkMode = false;
+
+    @override
+    void initState() { // initState Method: loads the theme preference from shared preference, when App starts
+        super.initState();
+        _loadTheme();
+    }
+    // _loadTheme Method: asynchronously loads preference from SharedPreferences, when App starts
+    _loadTheme() async {
```
Comment (biosfood) [inline theme-switch/main-theme-switch (V1.0).dart]: maybe just wait for this when booting?

Maybe profile this on some devices and only show the scene when everything is loaded

File: theme-switch/main-theme-switch (V1.0).dart [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    @override
+    Widget build(BuildContext context) {
+        return MaterialApp( // uses conditional to set theme to either dark or light, based on _isDarkMode booloan
+            theme: _isDarkMode ? ThemeData.dark() : ThemeData.light(),
+            home: Scaffold(
+                appBar: AppBar()
+                    title: Text('Theme Switch'),
+            ),
+            body: Center(
+                child: SwitchListTile( // Allows user to toggle theme, which updates the state and saves the preference
```
Comment (biosfood) [inline theme-switch/main-theme-switch (V1.0).dart]: Put this in a burger or side bar

File: theme-switch/main-theme-switch (V1.0).dart [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        _loadTheme();
+    }
+    // _loadTheme Method: asynchronously loads preference from SharedPreferences, when App starts
+    _loadTheme() async {
+        SharedPreferences prefs = await SharedPreferences.getInstance();
+        setState(() {
+            _isDarkMode = (prefs.getBool('isDarkMode') ?? false);
+        })
+    }
+    // _saveTheme Method: asynchronously saves theme preference to when changed
```
Comment (biosfood) [inline theme-switch/main-theme-switch (V1.0).dart]: spacing

File: theme-switch/main-theme-switch (V1.0).dart [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    _saveTheme(bool value) async {
+        _SharedPreferences prefs = await SharedPreferences.getInstance();
```
Comment (biosfood) [inline theme-switch/main-theme-switch (V1.0).dart]: use a global SharedPreferences object
This lets us avoid waiting here again

File: theme-switch/theme-switch-dependences.yml [hunk 11]
```diff
@@ -0,0 +1,7 @@
...
+  flutter:
+    sdk: flutter
+  shared_preference: ^2.0.13
```
Comment (biosfood) [inline theme-switch/theme-switch-dependences.yml]: Is this the right choice?
Some others maybe chime in here:
Is this enough for our storage needs or do we also need another storage?

**Review by biosfood — CHANGES_REQUESTED** [review]
Seems good in gerneral, I left some notes for the code style

File: theme-switch/main-theme-switch (V1.0).dart (hunk 1 above)
Comment (Lemonatix) [reply theme-switch/main-theme-switch (V1.0).dart]: Is removed now

File: theme-switch/main-theme-switch (V1.0).dart (hunk 5 above)
Comment (Lemonatix) [reply theme-switch/main-theme-switch (V1.0).dart]: main.dart is now consisting of the switch selector code

File: theme-switch/main-theme-switch (V1.0).dart (hunk 7 above)
Comment (Lemonatix) [reply theme-switch/main-theme-switch (V1.0).dart]: Still need to figure out, how to test it with apple device

File: theme-switch/main-theme-switch (V1.0).dart (hunk 8 above)
Comment (Lemonatix) [reply theme-switch/main-theme-switch (V1.0).dart]: Added Drawer (some kind of side bar)

File: theme-switch/main-theme-switch (V1.0).dart (hunk 10 above)
Comment (Lemonatix) [reply theme-switch/main-theme-switch (V1.0).dart]: Added a new ShapedPreferences object in main.dart

File: lib/main.dart [hunk 12]
```diff
@@ -1,20 +1,80 @@
...
+    void initState() {
+        super.initState();
+        _loadTheme();
   }
+
+    // Load theme preference from SharedPreferences
+    void _loadTheme() {
+        setState(() {
+            _isDarkMode = (prefs?.getBool('isDarkMode') ?? false);
+    });
```
Comment (biosfood) [inline lib/main.dart:30]: still an indentation problem here

File: lib/main.dart [hunk 13]
```diff
@@ -1,20 +1,80 @@
 import 'package:flutter/material.dart';
+import 'package:shared_preferences/shared_preferences.dart';
 
-void main() {
-  runApp(const MainApp());
+SharedPreferences? prefs;
```
Comment (biosfood) [inline lib/main.dart]: I would call this preferences

File: lib/main.dart [hunk 14]
```diff
@@ -1,20 +1,80 @@
...
+                                    color: Colors.blue,
+                                ),
+                            ),
+                            SwitchListTile(
+                                title: Text('Dark Mode'),
+                                value: _isDarkMode,
+                                onChanged: (value) {
+                                    setState(() {
+                                        _isDarkMode = value;
+                                         _saveTheme(value);
```
Comment (biosfood) [inline lib/main.dart]: indentation

File: lib/main.dart [hunk 15]
```diff
@@ -1,20 +1,80 @@
...
+    @override
+
+    Widget build(BuildContext context) {
+        return MaterialApp(
+            theme: _isDarkMode ? ThemeData.dark() : ThemeData.light(),
+            home: Scaffold(
+                appBar: AppBar(
+                    title: Text('Flutter Theme Switch'),
+                ),
+                drawer: Drawer(
```
Comment (biosfood) [inline lib/main.dart]: Great. I like it

File: pubspec.yaml [hunk 16]
```diff
@@ -9,6 +9,7 @@ environment:
 dependencies:
   flutter:
     sdk: flutter
+  shared_preference: ^2.0.13
```
Comment (biosfood) [inline pubspec.yaml]: shared_preferences

File: lib/main.dart [hunk 17]
```diff
@@ -1,20 +1,80 @@
 import 'package:flutter/material.dart';
+import 'package:shared_preferences/shared_preferences.dart';
 
-void main() {
-  runApp(const MainApp());
+SharedPreferences? prefs;
+
+void main() async {
+    WidgetsFlutterBinding.ensureInitialized();
+    prefs = await SharedPreferences.getInstance();
```
Comment (biosfood) [inline lib/main.dart]: I would prefer this to happen in the app but flutter seems to handle a loading screen quite nicely so this is okay for now

**Review by biosfood — CHANGES_REQUESTED** [review]
Organization is good now,
There are still a bunch of compiler warnings and some formatting errors that need fixing

Also, make sure to test everything

**Comment by biosfood** [comment]
Also, commits can and should consist of a single "action", you don't need a new commit for every single change.
Doesn't matter now, we can just squash everything in the end

File: pubspec.yaml (hunk 16 above)
Comment (biosfood) [reply pubspec.yaml]: looks good now

File: theme-switch/theme-switch-dependences.yml (hunk 11 above)
Comment (Lemonatix) [reply theme-switch/theme-switch-dependences.yml]: Going with this for now

**Review by biosfood — APPROVED** [review]
Looks good now.
Great job
:)

**Pull request merged**